### PR TITLE
MM-24036 Fixing some redirections and no teams handling on guest accounts

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Fixing the behavior of redirection when you are a guest and you are removed
from the last channel in your last team.

Small UI changes involved (added a try again text to provide a way to go back
to channels ones you are added again to the team).

#### Ticket Link
[MM-24036](https://mattermost.atlassian.net/browse/MM-24036)

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates

#### Device Information
This PR was tested on: Virtual Android device

#### Screenshots
image

### CODE REVIEW

**Comment by enahum** [comment]
@jespino build is failing

File: app/actions/views/channel.js [hunk 1]
```diff
@@ -200,18 +202,54 @@ export function selectDefaultChannel(teamId) {
...
+        const teamsMyFirstChannels = getMyFirstChannelForTeams(state);
+        for (const ch in Object.values(teamsMyFirstChannels)) {
+            if (ch) {
+                channelId = ch.id;
+                break;
+            }
         }
+
+        if (channelId) {
+            dispatch(handleSelectChannel(channelId));
```
Comment (enahum) [inline app/actions/views/channel.js:246]: don't you need to select the right team here too ?

**Comment by esethna** [comment]
@jespino can you please help provide screenshots of any UI you've changed?

File: app/actions/views/channel.js (hunk 1 above)
Comment (jespino) [reply app/actions/views/channel.js:246]: I don't think so, it is handled by the `handleSelectChannel` action it emits a `SELECT_CHANNEL` that includes the `teamId` extracted from the `channel` object itself.

Comment (enahum) [reply app/actions/views/channel.js:246]: What? HandleSelectChannel does not select a team

**Comment by jespino** [comment]
@esethna Done

File: app/actions/views/channel.js (hunk 1 above)
Comment (jespino) [reply app/actions/views/channel.js:246]: Ok, the handle selectChannel was passing the `team_id` parameter in the extra field of the action, and I was expecting was used to set the team too.. but yes, I think we should be selecting the team to there.

Comment (jespino) [reply app/actions/views/channel.js:246]: Fixed here f8d229710f47b97a68a9fef56ddce22fdd2a276d

**Comment by esethna** [comment]
Thanks JE. @michaelgamble have you reviewed this ticket? Also adding @matthewbirtch for a mobile UX review.

**Comment by jespino** [comment]
@enahum I'll take into consideration the DM/GM problem. Let me check it.

**Comment by jespino** [comment]
@enahum yes, you were totally right, now we only redirect you to other team if you are moving to a channel that is in another team (that means, have team_id value and the team_id is different to the current one).

**Comment by enahum** [comment]
@michaelgamble I encourage you to look at alternatives, what we have on desktop won't work well here without a good amount of refactoring. Also do look ahead as when having support for multiple servers.

I really doubt that the same flow as Desktop should be applied, the experience is totally different in my mind.

**Comment by michaelgamble** [comment]
> @michaelgamble I encourage you to look at alternatives, what we have on desktop won't work well here without a good amount of refactoring. Also do look ahead as when having support for multiple servers.
> 
> I really doubt that the same flow as Desktop should be applied, the experience is totally different in my mind.

why does it make sense to leave them on a blank screen with a message as opposed to logging them out with a message (keeping it consistent with what we do on desktop)?

**Comment by enahum** [comment]
I must add that there is no UI for displaying the message in the login screen as we have on the webapp and that will require design. If that is the direction we want to take, great, but I'll block this PR for going in into 1.31

**Comment by michaelgamble** [comment]
> I must add that there is no UI for displaying the message in the login screen as we have on the webapp and that will require design. If that is the direction we want to take, great, but I'll block this PR for going in into 1.31

Ah ok, that reason makes sense to me... ill ask Matt what he thinks real quick as he's been working on cleaning up the login flows.. brb on this PR

**Comment by enahum** [comment]
Looking at this a bit more, if you want to send the user to the `Login` screen, what happens with the Server Url ? the flow asks for the Server url screen to be the root screen and not login... do take that into consideration as well.

**Comment by jespino** [comment]
Yes, by the way, I think the current implementation for the webapp is not working like that, unless somebody else implemented it recently.

**Comment by michaelgamble** [comment]
Consulted with @matthewbirtch, we propose resolving this in two steps:
1) Band-aid solution by slight alteration to this PR
- remove the title "Select Team"
- remove "try again"

2) Design a proper fix for automatically loggin the user out and redirecting them to the login choice screen with message.
- created a UX design ticket to follow up on: https://mattermost.atlassian.net/browse/MM-24615

So ping me once step 1 is done, and then we can approve and close the PR and step 2 will trigger a new PR in a later release.

**Comment by jespino** [comment]
@michaelgamble removing the `try again` is pretty easy, but the problem is the user get trapped in that screen because there is no way to go anywhere else, except logging out. The `try again` is the solution to try to go back to the channels as soon as you are added to other channel.

**Comment by michaelgamble** [comment]
> @michaelgamble removing the `try again` is pretty easy, but the problem is the user get trapped in that screen because there is no way to go anywhere else, except logging out. The `try again` is the solution to try to go back to the channels as soon as you are added to other channel.

Well they aren't trapped, they have the "Log out" button as thier only option.. which is fine for now

**Comment by michaelgamble** [comment]
The only way a user ends up in this flow is if they have been removed from thier last channel as a guest.. so the reality of them sitting thier hitting "try again" over and over again while waiting for a channel admin to (in realtime) re-add them to a channel, i don't see that as a very likely use case

**Comment by jespino** [comment]
@michaelgamble I removed the "try again", remove the "Select Team" title is harder, do you have any suggestion to do that @enahum?

**Comment by enahum** [comment]
@jespino https://wix.github.io/react-native-navigation/api/options-api

Maybe @migbot can help you out, I’m currently unavailable

**Comment by jespino** [comment]
@migbot thanks, done

**Comment by amyblais** [comment]
@jespino Removing from v1.31 as we're past code complete. Let me know if you disagree and if this is actually urgent.

**Comment by jespino** [comment]
This is part of the Guest accounts stabilization and getting out of beta, probably @wiersgallak has a better understanding of the urgency of this.

**Comment by wiersgallak** [comment]
@amyblais @jespino This is fine if we can get into the next release.

**Comment by jespino** [comment]
Merged master

**Comment by josephbaylon** [comment]
@jespino @enahum 

The fix works on normal scenarios:
A.
1. Guest user logs in to app
2. Admin removes guest user from all channels

B.
1. Guest user is logged out
2. Admin removes guest user from all channels
3. Guest user logs in to app

Expected and Observed: User is redirected to this screen,
IMG_0114

However, in an edge case where
1. Guest user logs in to app
2. Admin removes guest user from all channels
3. Guest user is redirected to screen above
4. Admin adds user back to channel
5. Guest user closes and reopens app
6. Admin removes user again

Observed: Guest user is still on channel with the orange banner
IMG_0115

**Comment by josephbaylon** [comment]
@jespino @enahum Aside from issue above, there's unexpected error prompt when guest user is removed from all teams

1. Guest user is logged in
2. Admin removes guest user from all teams

IMG_0116

**Comment by enahum** [comment]
I huess this is conflicting with other fixes, the whole gust account thing seems to be very patchy and should be re-evaluated.

**Comment by michaelgamble** [comment]
@enahum I've created a ticket to reevaluate this and table a proper fix. https://mattermost.atlassian.net/browse/MM-25523

**Comment by josephbaylon** [comment]
@enahum @michaelgamble Should we remove this PR from v1.32 milestone? Also, this one https://mattermost.atlassian.net/browse/MM-24036

**Comment by enahum** [comment]
@josephbaylon I think so.. but that should be @amyblais call

**Comment by amyblais** [comment]
Sounds good to me, removing from v1.32,

**Comment by enahum** [comment]
@michaelgamble any progress with this? should @jespino close the PR and re-open when ready ?

**Comment by michaelgamble** [comment]
@enahum My suggestion would be to close it, and to create a new PR based on the solution of whatever comes out of the open jira ticket, but thats up to PM?

**Comment by enahum** [comment]
Not really up to PM.. closing.. feel free to re-open once this has been looked at

**Comment by michaelgamble** [comment]
sounds good :)

**Pull request closed without merging**